test/ce.integration.spec.ts

```diff
@@ -19,4 +20,91 @@ describe('ce.integration.spec.ts', () => {
     expect(host.textContent).toBe('One');
     dispose();
   });
+
+  it('should initialise web component when attached to document', async () => {
+    class CustomElementTwo {
+      static $view = '<template>One</template>';
+    }
+    const { host, registry, dispose } = await bootstrapAurelia({
+      root: class RootViewModel {
+        static $view = '<template></template>';
+      },
+      resources: [CustomElementTwo]
+    });
+    registry.register(CustomElementTwo);
+    await waitForTimeout(50);
+
+    const customComponent = document.createElement("custom-element-two");
+    host.appendChild(customComponent);
+    expect(customComponent.textContent).toBe('One');
+    dispose();
+  });
+
+  it('should not initialise web component when attribute changed on disconnected element', async () => {
+    class CustomElementThree {
+      static $view = '<template>One</template>';
+      // Bindable property must exist in order for the attributeChangedCallback to be executed
+      @bindable testing
+    }
+    const { registry, dispose } = await bootstrapAurelia({
+      root: class RootViewModel {
+        static $view = '<template></template>';
+      },
+      resources: [CustomElementThree]
+    });
+    registry.register(CustomElementThree);
+    await waitForTimeout(50);
+
+    const customComponent = document.createElement("custom-element-three");
+    customComponent.setAttribute("testing", "xx-ignored-xx");
+    expect(customComponent.textContent).toBe('One');
+    dispose();
+  });
+
+  it('should initialise web component with modified attribute value', async () => {
+    class CustomElementFour {
+      static $view = '<template>${testing}</template>';
+      // Bindable property must exist in order for the attributeChangedCallback to be executed
+      @bindable testing
+    }
+    const { host, registry, dispose } = await bootstrapAurelia({
+      root: class RootViewModel {
+        static $view = '<template></template>';
+      },
+      resources: [CustomElementFour]
+    });
+    registry.register(CustomElementFour);
+    await waitForTimeout(50);
+
+    const customComponent = document.createElement("custom-element-four");
+    customComponent.setAttribute("testing", "One");
+    host.appendChild(customComponent);
+    expect(customComponent.textContent).toBe('One');
```

Comment:
Is this test working for you? It reads like the `attributeChangedCallback` is getting fired immediately?

Reply:
You are correct in your reading, in that `attributeChangedCallback` is called before the `connectedCallback`. That this merely demonstrates that the `attributeChangedCallback` has no effect, as the behaviour does nothing until the controller is bound. At the point the controller is bound, it picks up the correct value for this attribute.

I'm definitely open to ways to clarify this better... it's really just exercising the various combinations of things and ensuring nothing bad happens 😺

Reply:
If so, can we do
```diff
    const customComponent = document.createElement("custom-element-four");
    customComponent.setAttribute("testing", "One");
+ expect(customComponent.textContent).toBe('');
    host.appendChild(customComponent);
    expect(customComponent.textContent).toBe('One');
```

Reply:
Ok... this assert now reveals something slightly odd. The `textContent` becomes `" "` (i.e., a single space).

That is:
1. Before the attribute is set, `textContent` is `""`
2. After the attribute is set, `textContent` is `" "`
3. After the element is attached, `textContent` is `"One"`

I don't have a good explanation for this. I've added the asserts here, and put a `// TODO` comment there to investigate. At least then you can review, and I'll dig into this further tomorrow perhaps...

Reply:
wow that's odd. A `// todo` will be fine, we will tackle it ... later :D